scylla/src/transport/query_result.rs

```diff
@@ -0,0 +1,654 @@
+use std::fmt::Debug;
+use std::marker::PhantomData;
+use std::sync::Arc;
+
+use thiserror::Error;
+use uuid::Uuid;
+
+use scylla_cql::frame::frame_errors::RowsParseError;
+use scylla_cql::frame::response::result::{
+    ColumnSpec, ColumnType, DeserializedMetadataAndRawRows, RawMetadataAndRawRows, RawRowsBorrowed,
+    RawRowsKind, RawRowsOwned, Row, TableSpec,
+};
+use scylla_cql::types::deserialize::result::TypedRowIterator;
+use scylla_cql::types::deserialize::row::DeserializeRow;
+use scylla_cql::types::deserialize::{DeserializationError, TypeCheckError};
+
+use super::legacy_query_result::LegacyQueryResult;
+
+/// A view over specification of a table in the database.
+#[derive(Debug, Clone, Copy)]
+#[cfg_attr(test, derive(PartialEq, Eq))]
+pub struct TableSpecView<'res> {
+    table_name: &'res str,
+    ks_name: &'res str,
+}
```

Comment:
The need for this is precisely caused by your requirement about moving TableSpecs back into ColumnSpec.

If you remember, I suggested that we move TableSpec out of ColumnSpec (as in the current implementation of Scylla and Cassandra it's always the same for all columns - no support for joins). You agreed only on condition that we will be able to revert this without introducing breaking changes once Scylla or Cassandra introduces joins.

`ColumnSpecView` will always borrow the keyspace name, no matter whether it's contained in `ResultMetadata` directly or in a `ColumnSpec`. Therefore, this abstraction layer defends us from the API breakage threat.
Conversely, if we returned `ColumnSpec` to the user instead of `ColumnSpecView`, then once we remove `TableSpec` from `ColumnSpec`, users will no longer be able to get table name and keyspace name from `ColumnSpec`, so this breaks the API.

Reply:
I don't see the benefit of removing TableSpec from ColumnSpec - it is borrowed anyway. Can we just keep it there and get rid of this additional layer?

Reply:
It is only borrowed (i.e., it contains `Cow::Borrowed` variants) if the metadata itself is borrowed as a whole, which is not possible for e.g. `RowIterator` (`query_iter()` API) and for `RowsDeserializer` in the owned variant. So, the additional layer saves possible allocations in such scenarios.

Reply:
This is no longer true now that metadata is always borrowed, right?

Reply:
Agreed. Apart from when being shared with PreparedStatement's cached metadata, result metadata is always deserialized in the borrowed form.

Reply:
I don't remember that very well. In case of errors we probably still want to have the table name stored there - but I'm not sure. We can merge this PR as-is, and discuss later (but before release!) if it's worth it to introduce the Views.